waspc/src/Analyzer/Parser/Lexer.x

```diff
@@ -0,0 +1,145 @@
+{
+-- This file is processed by Alex (https://www.haskell.org/alex/) and generates
+-- the module `Analyzer.Parser.Lexer`
+
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Analyzer.Parser.Lexer
+  ( lexer
+  ) where
+
+import Analyzer.Parser.Monad (ParserInput, Parser, ParserState (..), updatePosition, putInput, setStartCode)
+import Analyzer.Parser.Token (Token (..), TokenClass (..))
+import Analyzer.Parser.ParseError (ParseError (..))
+import Control.Monad.State.Lazy (get)
+import Control.Monad.Except (throwError)
+import Data.Word (Word8)
+import Codec.Binary.UTF8.String (encodeChar)
+}
+
+-- Character set aliases
+$digit = 0-9
+$alpha = [a-zA-Z]
+$identstart = [_$alpha]
+$ident = [_$alpha$digit]
+$any = [.$white]
+
+-- Regular expression aliases
+@string = \"([^\\\"]|\\.)*\" -- matches string-literal on a single line, from https://stackoverflow.com/a/9260547/3902376
+@double = "-"? $digit+ "." $digit+
+@integer = "-"? $digit+
+@ident = $identstart $ident* "'"*
+
+-- Tokenization rules (regex -> token)
+tokens :-
+
+-- Skips whitespace
+<0>       $white+ ;
+
+-- Quoter rules:
+-- Uses Alex start codes to lex quoted characters with different rules:
```

Comment:
Aha, so by entering <quoter> state, we are ignoring all the other rules, and the only thing we care about is hitting the first `tag=}`, once that happens other rules come back into life? Ok ok, makes sense!